Add API Key Metric for requests

AusIV opened the pull request:
We're going to start requiring keys for access to the gateway, and we want to know which keys are making which requests.

This adds the api key to prometheus metrics, and passes it through from the request.

Commit:
Add API Key Metric for requests

We're going to start requiring keys for access to the gateway,
and we want to know which keys are making which requests.

This adds the api key to prometheus metrics, and passes it through
from the request.

smoyer64 reviewed (changes requested):
The incoming `ApiKey` is a secret.  Instead of pushing that out into our observability DB, we should look up the identity of the user and tag with that instead.  I'm assuming the code to actually block the request isn't included since we need to wait until all users are sending their `ApiKey`.  Once we want to block invalid header values, we'll need a set containing those that are allowed.  We can use a map instead and the lookup can happen in the same step.  Want to add that code and comment out the blocking part?

Commit:
Don't forward API key headers

This moves the din api key from the header to the replacer, so that
sensitive information won't be forwarded to providers.

AusIV commented:
Hmmm... You're right. I hadn't thought about the fact that SigNoz is accessible to people we wouldn't want to have the raw API keys. 

At this point we're not implementing blocking because want visibility to who is providing the API key and who isn't, but also for the short term I was intending to implement blocking with off-the-shelf Caddy features, rather than building that into our middleware.

I'm also nervous at this stage to have to have a map configured while we're still trying to get adoption. If we find new teams that need keys, right now we can just make up a key for them, tell them to start using it, and start getting signoz data showing whether or not they are. If we have to have a map of keys to users, we'd have to update the map in the caddyfile and redeploy the router before we could see the data come into Signoz. That will become necessary later as we start actually requiring the keys, but right now that could be fairly frequent.

Let me ponder this problem for a bit.

Commit:
Translate prometheus api keys

This update removes private API keys from prometheus data, replacing
them with non-sensitive data.

If allows for a map of known users, and will send on a name that corresponds
to the API key when available.

Since we're not currently trying to block unknown users, but collect data
about unknown users, if a key isn't in the map we'll send signoz a 10
character salted hash of the API key, so we can issue keys and identify
when they're being used without having to redeploy caddy.

smoyer64 commented:
> If the key isn't in the map, we submit a salted hash of it.
This is a great interim behavior but once we "require" a Consensys-generated key, those should be manually put into the map (and no other keys should be accepted.)  If another key _does_ come in, we should definitely log that!

noleto reviewed (approved):
Looks good to me

smoyer64 commented:
Should we add a task to the backlog for the next step (denying unknown UUIDs)?
